MantidPlot/src/ProjectSerialiser.cpp

```diff
@@ -0,0 +1,948 @@
+  }
+  return true;
+}
+
+/**
+ * Serialise the state of Mantid
+ *
+ * This will go through all the parts that need to be serialised
+ * and create a string for them which will be written to the .mantid file.
+ *
+ * This will also save workspaces etc. to the project location.
+ *
+ * @param folder :: the folder to write out
+ * @return a string representation of the current project state
+ */
+QString ProjectSerialiser::serialiseProjectState(Folder *folder) {
+  QString text;
+
+  // Save the list of workspaces
+  if (window->mantidUI) {
+    text += saveWorkspaces();
+  }
+
+  // Save the scripting window
+  ScriptingWindow *scriptingWindow = window->getScriptWindowHandle();
+  if (scriptingWindow) {
+    std::string scriptString = scriptingWindow->saveToProject(window);
+    text += QString::fromStdString(scriptString);
+  }
+
+  // Finally, recursively save folders
+  if (folder) {
+    text += saveFolderState(folder, true);
+  }
+
+  return text;
+}
+
+/**
+ * Save the folder structure to a Mantid project file.
+ *
+ * @param app :: the current application window instance
+ * @return string represnetation of the folder's data
+ */
+QString ProjectSerialiser::saveFolderState(Folder *folder,
+                                           const bool isTopLevel) {
+  QString text;
+  bool isCurrentFolder = window->currentFolder() == folder;
+
+  if (!isTopLevel) {
+    text += saveFolderHeader(folder, isCurrentFolder);
+  }
+
+  text += saveFolderSubWindows(folder);
+
+  if (!isTopLevel) {
+    text += saveFolderFooter();
+  }
+
+  return text;
+}
+
+/**
+ * Generate the opening tags and meta information about
+ * a folder record for the Mantid project file.
+ *
+ * @param isCurrentFolder :: whether this folder is the current one.
+ * @return string representation of the folder's header data
+ */
+QString ProjectSerialiser::saveFolderHeader(Folder *folder,
+                                            bool isCurrentFolder) {
+  QString text;
+
+  // Write the folder opening tag
+  text += "<folder>\t" + QString(folder->objectName()) + "\t" +
+          folder->birthDate() + "\t" + folder->modificationDate();
+
+  // label it as current if necessary
+  if (isCurrentFolder) {
+    text += "\tcurrent";
+  }
+
+  text += "\n";
+  text += "<open>" + QString::number(folder->folderListItem()->isExpanded()) +
+          "</open>\n";
+  return text;
+}
+
+/**
+ * Generate the subfolder and subwindow records for the current folder.
+ * This method will recursively convert subfolders to their text representation
+ *
+ * @param app :: the current application window instance
+ * @param folder :: the folder to generate the text for.
+ * @param windowCount :: count of the number of windows
+ * @return string representation of the folder's subfolders
+ */
+QString ProjectSerialiser::saveFolderSubWindows(Folder *folder) {
+  QString text;
+
+  // Write windows
+  QList<MdiSubWindow *> windows = folder->windowsList();
+  foreach (MdiSubWindow *w, windows) {
```

Comment:
Can we use a range-based for-loop instead of the foreach macro? See [here](http://stackoverflow.com/questions/16825376/qt-foreach-loop-ordering-vs-for-loop-for-qlist?answertab=active#tab-top)